File: renderdocui/Windows/Dialogs/SuggestRemoteDialog.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using renderdocui.Code;

namespace renderdocui.Windows.Dialogs
{
    public partial class SuggestRemoteDialog : Form
    {
        public enum SuggestRemoteResult
        {
            Cancel,
            Local,
            Remote,
        }

        string m_WarningStart;

        public SuggestRemoteDialog(string driver, string machineIdent)
        {
            InitializeComponent();

            icon.Image = SystemIcons.Exclamation.ToBitmap();

            m_WarningStart =
                "This " + driver + " capture was originally created on a\n" +
                "'" + machineIdent.Trim() + "' machine.\n\n";

            warning.Text =
                m_WarningStart +
                "Currently you have no remote context selected or configured\n" +
                "to replay on. Would you like to load the capture locally or\n" +
                "back out to configure one in Tools > Manage Remote Servers?";

            remote.Enabled = false;
            remote.Image = null;
            remote.Text = "No Remote";
        }

        private void remoteDropDown_ItemAdded(object sender, ToolStripItemEventArgs e)
        {
            // update text and buttons to reflect that remote hosts are configured
            warning.Text =
                m_WarningStart +
                "Currently you have no remote context selected, would you like\n" +
                "to choose a remote context to replay on, or continue and load\n" +
                "the capture locally?";

            remote.Enabled = true;
            remote.Image = global::renderdocui.Properties.Resources.down_arrow;
            remote.Text = "Remote    ";
        }

        private void alwaysLocal_CheckedChanged(object sender, EventArgs e)
        {
            remote.Enabled = RemoteItems.Count > 0 && !alwaysLocal.Checked;
        }

        protected override bool ProcessCmdKey(ref Message msg, Keys keyData)
        {
            if (keyData == Keys.Escape)
            {
                this.Close();
                return true;
            }
            return base.ProcessCmdKey(ref msg, keyData);
        }

        private void cancel_Click(object sender, EventArgs e)
        {
            m_Result = SuggestRemoteResult.Cancel;
            Close();
        }

        private void local_Click(object sender, EventArgs e)
        {
            m_Result = SuggestRemoteResult.Local;
            Close();
        }

        private void remoteDropDown_ItemClicked(object sender, ToolStripItemClickedEventArgs e)
        {
            m_Result = SuggestRemoteResult.Remote;
        }

        private void remote_Click(object sender, EventArgs e)
        {
            if(remote.Checked)
                return;

            Point showPos = remote.PointToScreen(remote.ClientRectangle.Location);

            showPos.Y = showPos.Y + remote.ClientRectangle.Height;

            // gives the appearance that the button stays pressed
            remote.Checked = true;

            remoteDropDown.Show(showPos);
        }

        private void remoteDropDown_Closed(object sender, ToolStripDropDownClosedEventArgs e)
        {
            remote.Checked = false;

            // if the result is Remote now then one of the items was clicked on
            if (m_Result == SuggestRemoteResult.Remote)
                Close();
        }

        private SuggestRemoteResult m_Result = SuggestRemoteResult.Cancel;
        public SuggestRemoteResult Result
        {
            get
            {
                return m_Result;
            }
        }

        public ToolStripItemCollection RemoteItems
        {
            get
            {
                return remoteDropDown.Items;
            }
        }

        public bool AlwaysReplayLocally
        {
            get
            {
                return alwaysLocal.Checked;
            }
        }
    }
}

File: qrenderdoc/Windows/Dialogs/SuggestRemoteDialog.cpp
/******************************************************************************
 * The MIT License (MIT)
 *
 * Copyright (c) 2017-2019 Baldur Karlsson
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 ******************************************************************************/

#include "SuggestRemoteDialog.h"
#include <QMenu>
#include "Code/Resources.h"
#include "ui_SuggestRemoteDialog.h"

SuggestRemoteDialog::SuggestRemoteDialog(const QString &driver, const QString &machineIdent,
                                         QWidget *parent)
    : QDialog(parent), ui(new Ui::SuggestRemoteDialog)
{
  ui->setupUi(this);

  setWindowFlags(windowFlags() & ~Qt::WindowContextHelpButtonHint);

  m_WarningStart = tr("This %1 capture was originally created on a\n '%2' machine.\n\n")
                       .arg(driver)
                       .arg(machineIdent.trimmed());

  ui->warning->setText(m_WarningStart +
                       tr("Currently you have no remote context selected or configured\n") +
                       tr("to replay on. Would you like to load the capture locally or\n") +
                       tr("back out to configure one in Tools > Manage Remote Servers?"));

  ui->topFrame->setBackgroundRole(QPalette::Base);
  ui->topFrame->setAutoFillBackground(true);

  m_Remotes = new QMenu(this);

  ui->remote->setEnabled(false);
  ui->remote->setText(tr("No Remote"));

  ui->remote->setMenu(m_Remotes);

  QObject::connect(m_Remotes, &QMenu::triggered, this, &SuggestRemoteDialog::remoteItem_clicked);
}

SuggestRemoteDialog::~SuggestRemoteDialog()
{
  delete ui;
}

void SuggestRemoteDialog::remotesAdded()
{
  if(m_Remotes->isEmpty())
    return;

  // update text and buttons to reflect that remote hosts are configured
  ui->warning->setText(m_WarningStart +
                       tr("Currently you have no remote context selected, would you like\n") +
                       tr("to choose a remote context to replay on, or continue and load\n") +
                       tr("the capture locally?"));

  ui->remote->setEnabled(true);
  ui->remote->setText(tr("Remote"));
}

bool SuggestRemoteDialog::alwaysReplayLocally()
{
  return ui->alwaysLocal->isChecked();
}

void SuggestRemoteDialog::on_alwaysLocal_toggled(bool checked)
{
  ui->remote->setEnabled(!m_Remotes->isEmpty() && !checked);
}

void SuggestRemoteDialog::remoteItem_clicked(QAction *action)
{
  m_Choice = Remote;
  accept();
}

void SuggestRemoteDialog::on_local_clicked()
{
  m_Choice = Local;
  accept();
}

void SuggestRemoteDialog::on_cancel_clicked()
{
  m_Choice = Cancel;
  accept();
}
